File: C++/bt4.cpp
#include <iostream>
#include <vector>
#include <chrono>
using namespace std;

const int INSERTION_SORT_THRESHOLD = 10;

// Hàm sắp xếp bằng Insertion Sort
void insertionSort(vector<int>& arr, int low, int high) {
    for (int i = low + 1; i <= high; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= low && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

// Hàm phân vùng cho Quick Sort
int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

// Hàm cài đặt Hybrid Quick Sort
void hybridQuickSort(vector<int>& arr, int low, int high) {
    if (high - low < INSERTION_SORT_THRESHOLD) {
        insertionSort(arr, low, high);
    } else {
        int pi = partition(arr, low, high);
        hybridQuickSort(arr, low, pi - 1);
        hybridQuickSort(arr, pi + 1, high);
    }
}

int main() {
    vector<int> sizes = {5000, 10000, 20000};
    for (int size : sizes) {
        vector<int> arr(size);
        for (int i = 0; i < size; i++)
            arr[i] = rand() % 10000;

        auto start = chrono::high_resolution_clock::now();
        hybridQuickSort(arr, 0, size - 1);
        auto end = chrono::high_resolution_clock::now();
        chrono::duration<double, milli> duration = end - start;
        cout << "Thời gian Hybrid Quick Sort với kích thước " << size << ": " << duration.count() << " ms" << endl;
    }
    return 0;
}

File: python/bt4.py
import random
import time

INSERTION_SORT_THRESHOLD = 10

# Hàm sắp xếp bằng Insertion Sort
def insertion_sort(arr, low, high):
    for i in range(low + 1, high + 1):
        key = arr[i]
        j = i - 1
        while j >= low and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

# Hàm phân vùng cho Quick Sort
def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

# Hàm cài đặt Hybrid Quick Sort
def hybrid_quick_sort(arr, low, high):
    if high - low < INSERTION_SORT_THRESHOLD:
        insertion_sort(arr, low, high)
    else:
        pi = partition(arr, low, high)
        hybrid_quick_sort(arr, low, pi - 1)
        hybrid_quick_sort(arr, pi + 1, high)

# Chạy thử nghiệm
sizes = [5000, 10000, 20000]
for size in sizes:
    arr = [random.randint(0, 10000) for _ in range(size)]

    start_time = time.time()
    hybrid_quick_sort(arr, 0, len(arr) - 1)
    hybrid_duration = (time.time() - start_time) * 1000
    print(f"Thời gian Hybrid Quick Sort với kích thước {size}: {hybrid_duration:.2f} ms")